### PR TITLE
Fix synchronized memcpy in GPT

### PR DESCRIPTION
**Pull request by Wong4j**
### PR types
Performance optimization

### PR changes
Others

### Description
Avoid synchronized memcpy in GPT pretraining

### CODE REVIEW

File: paddlenlp/transformers/gpt/modeling.py [hunk 1]
```diff
@@ -1240,8 +1240,12 @@ def forward(self, prediction_scores, masked_lm_labels, loss_mask=None):
         """
         with paddle.amp.auto_cast(False):
             masked_lm_loss = self.loss_func(prediction_scores.astype("float32"), masked_lm_labels.unsqueeze(2))
-            masked_lm_loss = masked_lm_loss[masked_lm_loss > 0].astype("float32")
-            loss = paddle.mean(masked_lm_loss)
+            # skip ignore_index which loss == 0
+            if loss_mask is None:
+                loss_mask = (masked_lm_loss > 0).astype("float32")
+                loss_mask = loss_mask.reshape([-1])
+            masked_lm_loss = paddle.sum(masked_lm_loss.reshape([-1]) * loss_mask)
```
Comment (ZHUI) [inline paddlenlp/transformers/gpt/modeling.py:1247]: Support for using custom loss mask?

Comment (Wong4j) [reply paddlenlp/transformers/gpt/modeling.py:1247]: 这里的改动主要是因为 `masked_lm_loss[masked_lm_loss > 0]` 的写法会导致D2H的copy。改成loss_mask与lm_loss相乘，得到masked_lm_loss，两种是等效的，但不会有D2H copy。

Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling.py:1247]: 是 slice op的原因吗？

Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling.py:1247]: 这里可能需要注意下 最后 masked_lm_loss  的数据类型，希望是 float32的

File: paddlenlp/transformers/gpt/modeling.py [hunk 2]
```diff
@@ -1240,8 +1240,12 @@ def forward(self, prediction_scores, masked_lm_labels, loss_mask=None):
         """
         with paddle.amp.auto_cast(False):
             masked_lm_loss = self.loss_func(prediction_scores.astype("float32"), masked_lm_labels.unsqueeze(2))
-            masked_lm_loss = masked_lm_loss[masked_lm_loss > 0].astype("float32")
```
Comment (Xreki) [inline paddlenlp/transformers/gpt/modeling.py:1243]: 返回的`masked_lm_loss`的shape跟`masked_lm_loss > 0`比较结果中True的个数有关，因此需要把`masked_lm_loss > 0`比较结果中True的个数传回CPU，因此需要一个DtoH拷贝。`masked_lm_loss[masked_lm_loss > 0]`的实现无法避免这个`DtoH`的。PR中的修改避开了getitem操作，实现了同样的功能，并且避免了DtoH拷贝。

File: paddlenlp/transformers/gpt/modeling.py (hunk 1 above)
Comment (Xreki) [reply paddlenlp/transformers/gpt/modeling.py:1247]: 原实现`masked_lm_loss = masked_lm_loss[masked_lm_loss > 0].astype("float32")`，返回的`masked_lm_loss`的shape跟`masked_lm_loss > 0`比较结果中True的个数有关，因此需要把`masked_lm_loss > 0`比较结果中True的个数传回CPU，因此需要一个DtoH拷贝。`masked_lm_loss[masked_lm_loss > 0]`的实现无法避免这个`DtoH`的。

PR中的修改避开了getitem操作，实现了同样的功能，并且避免了DtoH拷贝。

Comment (Wong4j) [reply paddlenlp/transformers/gpt/modeling.py:1247]: > 这里可能需要注意下 最后 masked_lm_loss 的数据类型，希望是 float32的

现在的写法应该可以确保是float32吧？

**Review by ZHUI — APPROVED** [review]
LGTM
